sourcegraph/sourcegraph
# remove limited gql backend

Related to https://github.com/sourcegraph/sourcegraph/issues/41145

Alright, what's the big idea?

### The big idea

This PR is "simply" removing `CodeInsightsGqlBackendLimited`. To do that we:

1. Create a new Zustand store to hold our license info
2. Move license check into a new hook that saves results to the store
3. Update `useApi` to always return the full `CodeInsightsGqlBackend`
4. Update `useUiFeatures` to reference the store
5. Find all references of `CodeInsightsGqlBackendContext.UiFeatures` and update them to use `useUiFeatures`
6. Delete `useUiFeatures` tests. (see explanation below)
7. Add integration tests
8. Update jest tests
9. Update storybook stories

### Ok, and this helps us....how exactly?

By itself, this PR doesn't do much (other than add some much-needed test coverage). But it isolates the primary logic change from the next PR that will remove the Context and Provider. That will involve a lot of boilerplate that would muddy up the logic changes here. Removing the context gives us our big win. It will make refactoring the API **much** easier and makes navigating the code actually possible. i.e. cmd+click a method definition takes you to the definition, not the context placeholder.

### Footnotes

6. I deleted the test because right now, it only asserts `licensed` correctly references the appropriate backend. Since we only have one backend that check no longer makes sense. The updated test would just assert that the store saves itself, which is meaningless. I considered updating the test to check the dynamic functions, but that would be out of scope for this PR. I also considered leaving the test in place but empty, but that feels even more wrong. I suggest covering `useUiFeatures` implicitly with integration tests that consume it.

> Screenshot notes: the before and after screenshots should be identical for their respective license/unlicensed modes

### Before screenshots

#### Licensed

![PixelSnap 2022-10-05 at 09 52 30](https://user-images.githubusercontent.com/1855233/194091761-27717e4f-750a-4ad6-a0e7-46af470411f0.png)

#### Unlicensed

![PixelSnap 2022-10-05 at 09 53 32](https://user-images.githubusercontent.com/1855233/194091926-912895b3-033b-4927-a844-d500f96647e2.png)

### After screenshots

#### Licensed

![PixelSnap 2022-10-05 at 09 43 21](https://user-images.githubusercontent.com/1855233/194089593-5ea74527-390d-45a0-854d-dcb0625571d3.png)

#### Unlicensed

![PixelSnap 2022-10-05 at 09 47 50](https://user-images.githubusercontent.com/1855233/194090563-61b17368-6293-4ad6-805e-6153ca62d236.png)

## Test plan

A note on real vs fake testing. To test this locally you can "fake" the license quickly by just updating the license check in `use-license` by setting `licensed` to `true` or `false`.

This works for testing most of the UI, but the locked insight cards are only displayed if the license is actually updated (because that check happens on the backend).

If you want to perform a "real" test, you need to [update your license accordingly](https://docs.sourcegraph.com/dev/how-to/generate_license_key_for_testing).

- Integration tests should pass
- Set `licensed: true` 
- The licensed instance should render the dashboard without limited-mode banners.
- Set `licensed: false`
- The unlicensed instance should render the dashboard with limited-mode banners.

## App preview:

- [Web](https://sg-web-insights-remove-limited-backend.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-hlmgojwooj.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
